[port/1.21] Fix issues relating to gameplay-enchantment support

Shadows-of-Fire opened the pull request:
The switch to data enchantments has left `GetEnchantmentLevelEvent` and `IItemExtension#getAllEnchantments` inoperable due to the inability to resolve `Holder`s, which are required to populate an `ItemEnchantments` instance.  This PR provides the necessary `RegistryLookup<Enchantment>` to these contexts, retrieving it "as safely as possible".

The term "as safely as possible" means compromising between accessing magic global state and patching a ton of locations.  The primary technique used to access the `RegistryLookup` is by retrieving it from the existing `Holder.Reference` objects, or whatever other context is available (such as an entity).

In two cases, there was no relevant context available, which means that the global state has to be accessed through the new `CommonHooks.resolveLookup` function. The cases were:
1. `ItemEnchantmentsPredicate.Enchantments#matches`, which only has `ItemStack` context, and
2. `EnchantmentHelper#runIterationOnItem(ItemStack, EnchantmentHelper.EnchantmentVisitor)`, which does not have any access to a server or level.

All other cases use available local state to resolve the `RegistryLookup`.

In addition, this PR resolves a todo in `ArrowItem#isInfinite`, updating it to reflect current systems.

Shadows-of-Fire committed:
Fix issues relating to gameplay-enchantment support

Shadows-of-Fire committed:
sptles

Shadows-of-Fire committed:
Fix enchantment level tests

Shadows-of-Fire committed:
Clarify separation of gameplay and nbt enchants

pupnewfster commented on patches/net/minecraft/advancements/critereon/ItemEnchantmentsPredicate.java.patch:
```diff
@@ -1,14 +1,19 @@
-+            return matches(p_333958_, p_333958_.getAllEnchantments());
++            var lookup = net.neoforged.neoforge.common.CommonHooks.resolveLookup(net.minecraft.core.registries.Registries.ENCHANTMENT);
++            if (lookup != null) {
++                matches(p_333958_, p_333958_.getAllEnchantments(lookup));
```
Shouldn't this have a `return`?

pupnewfster commented on patches/net/minecraft/core/Holder.java.patch:
```diff
@@ -21,6 +23,18 @@
++     * This will only succeed if the underlying holder is a {@link Holder.Reference}.
++     */
++    @org.jetbrains.annotations.Nullable
++    default HolderLookup.RegistryLookup<T> unwrapLookup() {
```
Given we are already patching the interface to make it extend `IWithData`, do we want to use have an `IHolderExtension` for these methods rather than us patching in extra methods onto `Holder` itself?

pupnewfster commented on patches/net/minecraft/world/item/ArrowItem.java.patch:
```diff
@@ -1,13 +1,21 @@
++     * @param ammo The ammo stack (containing this item)
++     * @param bow  The bow stack
++     * @param livingEntity The entity who is firing the bow
++     * @return True if the arrow is infinite
```
We should probably specify the behavior in the docs that unlike what we did in older mc versions, this doesn't return true by default for if the bow has the infinity enchant.

Though thinking about it, does not defaulting it like we used to mean that it is not possible to make an arrow not be infinite regardless of the enchantments on the bow?

Shadows-of-Fire replied:
It does, yeah.  Delegating that full control to the `ArrowItem` will require updating the method to basically just be a call wrapper for `EnchantmentHelper.processAmmoUse`

pupnewfster commented on patches/net/minecraft/world/item/ProjectileWeaponItem.java.patch:
```diff
@@ -9,12 +9,13 @@
 -        int i = !p_330934_ && !p_330302_.hasInfiniteMaterials() && p_330302_.level() instanceof ServerLevel serverlevel
-+        int i = !p_330934_ && !(p_330302_.hasInfiniteMaterials() || (p_331434_.getItem() instanceof ArrowItem && ((ArrowItem) p_331434_.getItem()).isInfinite(p_331434_, p_331207_, p_330302_))) && p_330302_.level() instanceof ServerLevel serverlevel
++        // Neo: Adjust this check to respect ArrowItem#isInfinite, bypassing processAmmoUse if true.
++        int i = !p_330934_ && p_330302_.level() instanceof ServerLevel serverlevel && !(p_330302_.hasInfiniteMaterials() || (p_331434_.getItem() instanceof ArrowItem && ((ArrowItem) p_331434_.getItem()).isInfinite(p_331434_, p_331207_, p_330302_)))
```
This patch would be more readable if we made use of newer java language features and patterned matched the cast:
```java
p_331434_.getItem() instanceof ArrowItem arrowItem && arrowItem.isInfinite(p_331434_, p_331207_, p_330302_)
```
A separate thought is do we want to be patching the method onto `ArrowItem`, or do we want to make some sort of projectile item interface that contains it, so that modders can implement it on their own projectiles to potentially make having them be infinite easier?

pupnewfster commented on src/main/java/net/neoforged/neoforge/client/ClientHooks.java:
```diff
@@ -1048,4 +1051,14 @@ public static void fireClientTickPre() {
     }
+
+    @Nullable
+    @SuppressWarnings("resource")
```
My IDE is telling me this is a redundant suppression

Shadows-of-Fire replied:
Eclipse gets mad if I remove it, granted, I think eclipse is stupid here, but still: 
 ![](https://i.imgur.com/yIAEaBW.png)

pupnewfster replied:
Oh that’s what that suppression is for… I just have that off globally in my IDE given it is annoying for when using levels

pupnewfster commented on src/main/java/net/neoforged/neoforge/common/extensions/IItemExtension.java:
```diff
@@ -457,13 +458,14 @@ default int getEnchantmentLevel(ItemStack stack, Holder<Enchantment> enchantment
     @ApiStatus.OverrideOnly
-    default ItemEnchantments getAllEnchantments(ItemStack stack) {
+    default ItemEnchantments getAllEnchantments(ItemStack stack, RegistryLookup<Enchantment> lookup) {
         return stack.getOrDefault(DataComponents.ENCHANTMENTS, ItemEnchantments.EMPTY);
```
I think it might be slightly clearer what this is doing if we have the default implementation call `ItemStack#getTagEnchantments`

pupnewfster commented on patches/net/minecraft/world/item/ItemStack.java.patch:
```diff
@@ -140,6 +140,28 @@
++     * Gets all enchantments from NBT. Use {@link ItemStack#getAllEnchantments} for gameplay logic.
++     */
++    public ItemEnchantments getTagEnchantments() {
++        return this.getOrDefault(DataComponents.ENCHANTMENTS, ItemEnchantments.EMPTY);
```
Do we want to default this to calling `getEnchantments` rather than duplicating the code inside of it? Also could you move this method above `getEnchantments` so that it is a single chunk of stuff being patched in (even though I am aware that regardless it still is just one hunk)

pupnewfster commented on src/main/java/net/neoforged/neoforge/common/extensions/IItemExtension.java:
```diff
@@ -457,13 +458,14 @@ default int getEnchantmentLevel(ItemStack stack, Holder<Enchantment> enchantment
      */
     @ApiStatus.OverrideOnly
-    default ItemEnchantments getAllEnchantments(ItemStack stack) {
+    default ItemEnchantments getAllEnchantments(ItemStack stack, RegistryLookup<Enchantment> lookup) {
```
Please fix the javadoc links in this method and in `getEnchantmentLevel`

pupnewfster commented on src/main/java/net/neoforged/neoforge/common/extensions/IItemStackExtension.java:
```diff
@@ -155,16 +156,13 @@ default boolean canApplyAtEnchantingTable(Enchantment enchantment) {
      * Gets the gameplay level of the target enchantment on this stack.
      * <p>
-     * Equivalent to calling {@link EnchantmentHelper#getItemEnchantmentLevel(Enchantment, ItemStack)}.
+     * Use in place of {@link EnchantmentHelper#getTagEnchantmentLevel(Enchantment, ItemStack)} for gameplay logic.
```
Broken javadoc link, the first parameter should be a `Holder` not `Enchantment`

pupnewfster commented on src/main/java/net/neoforged/neoforge/event/EventHooks.java:
```diff
@@ -1038,8 +1039,11 @@ public static boolean onEffectRemoved(LivingEntity entity, MobEffectInstance eff
-        var event = new GetEnchantmentLevelEvent(stack, enchantments, ench);
-        NeoForge.EVENT_BUS.post(event);
+        RegistryLookup<Enchantment> lookup = ench.unwrapLookup();
+        if (lookup != null) { // Pretty sure this is never null, but I can't *prove* that it isn't.
```
Can we restructure the body of this event to avoid the extra instantiation in case lookup does end up being null? For example:
```java
RegistryLookup<Enchantment> lookup = ench.unwrapLookup();
if (lookup == null) { // Pretty sure this is never null, but I can't *prove* that it isn't.
    return level;
}
var enchantments = new ItemEnchantments.Mutable(ItemEnchantments.EMPTY);
enchantments.set(ench, level);
var event = new GetEnchantmentLevelEvent(stack, enchantments, ench, lookup);
NeoForge.EVENT_BUS.post(event);
return enchantments.getLevel(ench);
```

pupnewfster commented on patches/net/minecraft/world/item/enchantment/EnchantmentHelper.java.patch:
```diff
@@ -1,30 +1,46 @@
      public static int getItemEnchantmentLevel(Holder<Enchantment> p_346179_, ItemStack p_44845_) {
--        ItemEnchantments itemenchantments = p_44845_.getOrDefault(DataComponents.ENCHANTMENTS, ItemEnchantments.EMPTY);
--        return itemenchantments.getLevel(p_346179_);
++        // To reduce patch size, update this method to always check gameplay enchantments, and add getTagEnchantmentLevel as a helper for mods.
```
Could we get the `Neo: ` prefix added to this comment?

pupnewfster commented on patches/net/minecraft/world/item/enchantment/EnchantmentHelper.java.patch:
```diff
@@ -1,30 +1,46 @@
++        // Neo: Respect gameplay-only enchantments when doing iterations
++        var lookup = net.neoforged.neoforge.common.CommonHooks.resolveLookup(net.minecraft.core.registries.Registries.ENCHANTMENT);
++        if (lookup != null) {
++            itemenchantments = p_345425_.getAllEnchantments(lookup);
```
Do we want to pass in `ItemEnchantments` to the method to avoid having to look them up an extra time in the item extension from the stack? (I don't really care one way or another, so feel free to resolve this after thinking about it)

Shadows-of-Fire replied:
That sounds like a lot of change for basically no benefit, retrieving the enchantments from the stack is effectively free.

pupnewfster reviewed (commented):
Left some comments, it also seems like `IItemStackExtension#canApplyAtEnchantingTable(Enchantment)` is currently a dead method? Is it still a possible hook? If not we should probably remove it. If it is a possible hook, then we likely want to hook it back up and change it to take a `Holder<Enchantment>` instead  of an `Enchantment`

We also will need to make sure in the neo docs (cc: @ChampionAsh5357 ) that we specify the behavior of certain `EnchantmentHelper` methods compared to how we patch things. For example: `EnchantmentHelper#hasAnyEnchantments` just checks about tag based enchantments and not gameplay ones (which is correct from the point of view that it is only used by the grindstone in vanilla, but naming wise might confuse people).

One other thought is do we want to be patching `EnchantmentHelper#hasTag` to check gameplay enchantments? The uses I see are:
- Beehive block, used to check if the stack is enchanted with an enchantment in `EnchantmentTags.PREVENTS_BEE_SPAWNS_WHEN_MINING`
- Decorated pot, used to check if the stack is enchanted with an enchantment in `EnchantmentTags.PREVENTS_DECORATED_POT_SHATTERING`
- Ice block, used to check if the stack is enchanted with an enchantment in `EnchantmentTags.PREVENTS_ICE_MELTING`
- Infested blocks, used to check if the stack is enchanted with an enchantment in `EnchantmentTags.PREVENTS_INFESTED_SPAWNS`

I believe all of those are gameplay effects, so I think it would make sense for us to patch the helper method to be able to check our gameplay only ones.

Shadows-of-Fire committed:
Add `isPrimaryItemFor` hook

This hook replaces `canApplyAtEnchantingTable` with a similar purpose, but adapted to the name changes made by vanilla.

Shadows-of-Fire commented:
`IItemExtension#canApplyAtEnchantingTable` was reimplemented as `IItemExtension#isPrimaryItemFor(Holder<Enchantment>)`

Shadows-of-Fire committed:
Fix missing return in ItemEnchantmentsPredicate

Shadows-of-Fire committed:
Move to IHolderExtension

Shadows-of-Fire committed:
address various review comments

Shadows-of-Fire committed:
spotless

Shadows-of-Fire committed:
Patch in `getAllEnchantments` to `EnchantmentHelper#hasTag`
